### DISCUSSION TITLE
Task 3: Fill in missing data

### DISCUSSION THREAD

**Discussion by rsbgonzalez** [rramosp/ai4eng.v1#112] [02 - PYTHON]
Buenas tardes para @rramosp, tengo problemas con el código para esta tarea.

Estoy utilizando la función fillna, pero no me está reemplazando los valores Nan con los valores que muestreo, el resultado es "None" si lo hago convirtiendo el array de muestreo a una serie de pandas o a un data frame, para ambos casos utilizando inplace = True.

Sin embargo, cuando no defino el inplace, el resultado es el dataframe original

Este es mi código 
copia=d.copy()
pm = d.mean()['price']
pstd = d.std()['price']
Nan = len(d) - d.count()[0]
indice = np.arange(1,len(d)+1)
sample = np.random.normal(pm,pstd,len(d))
samples = pd.DataFrame(sample, index = indice,  columns=["price"])
#samples = pd.Series(sample)
result = d.fillna(samples,inplace = True)
print(result)

Agradezco su colaboración.

**Comment by Felipe-RA** [comment]
Saludos rsbgonzales

No comprendo para que creaste la variable Nan, no ve que le des uso.

Para calcular el número de elementos nan en una columna te recomiendo utilizar `column.isna().sum()` en donde `column` es la columna en la que quieres contar el número de nan.

Ademas, el objetivo de crear un `copy` del Dataframe es no tener que construir un dataframe nuevo desde cero, que es lo que estas haciendo al crear un dataframe nuevo al crear `samples`.

Por ultimo, para entender por qué  el `fillna` a veces te devuelve null vamos lo podemos encontrar en [la documentación](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html)

![image](https://user-images.githubusercontent.com/49454068/153026305-84886dbc-75e5-4cb1-ba90-188ab05ba59f.png)

Entonces, si haces un inplace TRUE y le asignas ese valor a una variable utilizando fillna estas retornando None.

Y cuando implementas el inplace FALSE en la función (que es para cambiar los valores sobre la marcha en el dataframe) te da el original porque no estas utilizando la copia ni le estas pasando la normal que calculaste a la función fillna.

Fijate en los parametros que recibe la función fillna() (ESPECIAL ATENCIÓN AL `value` que es lo que te falta)
![image](https://user-images.githubusercontent.com/49454068/153027110-16fb88cf-dbe7-4fb5-b4d4-3539183d257c.png)

**Reply by Felipe-RA** [reply]
Cualquier cosa me comentas

Saludos

**Reply by rsbgonzalez** [reply]
Hola Felipe, muchas garcias por tu respuesta.

Cambié mi código con la sugerencia que hizo el profe en otro post, pero tampoco me funcionó, me arroja el mismo dataframe con los Nan:

Respecto a samples, es el vector que contiene los números aleatorios que se muestrearon convertido a dataframe


pm = d.mean()['price']
pstd = d.std()['price']
size= d["price"].isna().sum()+1
indice = np.arange(1,size)
sample = np.random.normal(pm,pstd,size=d["price"].isna().sum())
samples = pd.Series(sample)
d["price"][d.price.isna()] = samples
result = d
print(result)

Agradezco mucho tu ayuda

**Comment by rsbgonzalez** [comment]
Hola Felipe, muchas garcias por tu respuesta.

Cambié mi código con la sugerencia que hizo el profe en otro post, pero
tampoco me funcionó, me arroja el mismo dataframe con los Nan:

Respecto a samples, es el vector que contiene los números aleatorios que se
muestrearon convertido a dataframe


pm = d.mean()['price']
pstd = d.std()['price']
size= d["price"].isna().sum()+1
indice = np.arange(1,size)
sample = np.random.normal(pm,pstd,size=d["price"].isna().sum())
samples = pd.Series(sample)
d["price"][d.price.isna()] = samples
result = d
print(result)

Agradezco mucho tu ayuda

Ricardo Sebastián González
BSc. Chemical Engineering
MSc. Renewable Energy - Concentrated Solar Energy Specialist
Mobile:+573136523653 - Colombia

**Reply by Felipe-RA** [reply]
Hay una diferencia importante entre `sample` y `samples`

En tu código original: `sample` es el vector con los números muestreados 
`samples`  es el dataframe nuevo que creaste

Ahora en el nuevo código que me mandaste `samples` es una serie de pandas con los valores muestreados.

También fijate que te están pidiendo que no modifiques el dataframe original, sino que crees una copia.

Por ultimo, vamos a ver esta linea de código:

`d["price"][d.price.isna()] = samples`

Recuerda que no deberias estar trabajando sobre el dataframe original d, sino sobre la copia.

Una vez tengas en cuenta eso puedes hacer lo siguiente:

`copy['price'].loc[ copy['price'].isna() == True ] = samples` 

Fijate que `.loc[ copy['price'].isna() == True ]` reemplaza tu `d.price.isna()]`

FInalmente, te recomiendo que limpies tu código de variables que ya no estas utilizando (index) y utilices nombres menos parecidos para no confundirte (entre sample y samples por ejemplo)

Cualquier cosa me dices, hasta pronto

**Reply by rsbgonzalez** [reply]
Hola Felipe,

Así me quedó el código con tus recomendaciones, pero... sigo obteniendo el dataframe sin reemplazar los Nan

copia=d.copy()
pm = copia.mean()['price']
pstd = copia.std()['price']
size= copia["price"].isna().sum()+1
sample = np.random.normal(pm,pstd,size)
samples = pd.Series(sample)
copia["price"].loc[copia["price"].isna()== True] = samples
copia

**Comment by rsbgonzalez** [comment]
Felipe, sólo para entender mejor el tema del fillna().

Por qué retorna None cuando se asigna inplace= True?. Por que mira que en este caso que hay en las notas, no está dando como output None:


![Captura de pantalla 2022-02-08 114759](https://user-images.githubusercontent.com/85143626/153035071-964dfe64-7937-4518-970e-6e4dcdeb4aad.jpg)

**Reply by Felipe-RA** [reply]
En este caso `fillna` no retorna None porque es una operación que se esta aplicando **sobre el `Dataframe`**

Mira este ejemplo:

`input`
```
...
...

d.Brebeuf.fillna(values_to_add. inplace= True)
print(type(d))
```

`Output`
```
pandas.DataFrame
```

Ahora mira este otro ejemplo con inplace= True

`input`
```
v = d.Brebeuf.fillna(values_to_add. inplace= True)
print(type(v)
```
`Output`

```
None
```

La diferencia es que en el primer caso **estas aplicando una función sobre el dataframe `d`** mientras que en el segundo caso **estas asignandole el valor a `v` del resultado de aplicarle una función al dataframe `d`**

**Reply by rsbgonzalez** [reply]
Muy bien, ahora lo entiendo perfectamente, muchas gracias!

Me ayudas porfa con mi código?, esta es la última versión, pero sigue arrojandome el dataframe sin llenar los Nan:

copia=d.copy()
pm = copia.mean()['price']
pstd = copia.std()['price']
size= copia["price"].isna().sum()+1
indice = np.arange(1,size+1)
sample = np.random.normal(pm,pstd,size)
#samples = pd.DataFrame(sample, index = indice,  columns=["price"])
samples = pd.Series(sample)
copia.fillna(samples)
copia
